substrate/client/network/src/service.rs

```diff
@@ -1456,11 +1451,17 @@ where
 					listen_addrs.truncate(30);
 				}
 				for addr in listen_addrs {
-					self.network_service
-						.behaviour_mut()
-						.add_self_reported_address_to_dht(&peer_id, &protocols, addr);
+					self.network_service.behaviour_mut().add_self_reported_address_to_dht(
+						&peer_id,
+						&protocols,
+						addr.clone(),
+					);
 				}
 				self.peer_store_handle.add_known_peer(peer_id);
+				// Confirm the observed address manually since they are no longer trusted by
+				// default (libp2p >= 0.52)
+				// TODO: remove this when/if AutoNAT is implemented.
+				self.network_service.add_external_address(observed_addr);
```

Comment:
With this addition every observed address reported through `libp2p-identify` is added as an external address, i.e. is being confirmed as an external address.

Currently you are not using [TCP port-reuse](https://docs.rs/libp2p-tcp/latest/libp2p_tcp/struct.Config.html#method.port_reuse), thus the TCP transport uses a new port for every outgoing connection. Thus each remote peer on outgoing connections observes the local node under a different port. Given that these outgoing connection ports are ephemeral, none of them are a suitable external address candidate.

When not using TCP port-reuse, one can replace the observed port of an outgoing connection with a listening port, in the hope that one can thus construct a suitable external address. This is what `libp2p-swarm` is doing by default before passing a reported candidate to each of the `NetworkBehaviour` implementations. The [process is called `address_translation`](https://docs.rs/libp2p-core/0.40.1/libp2p_core/transport/trait.Transport.html#tymethod.address_translation). 

https://github.com/libp2p/rust-libp2p/blob/1e3ffeb7990e23660e30dd95fbe48c1a7138f102/swarm/src/lib.rs#L1092-L1123

The problem is that the line above uses the vanilla observed addresses from `libp2p-identify`, i.e. side-stepping the `address_translation` mechanism. Thus the line is confirming each ephemeral outgoing connection port as a valid external address. That should be the reason why you are seeing so many external addresses reported in https://github.com/libp2p/rust-libp2p/discussions/4688.

I see multiple paths forward:

- Only call `self.network_service.add_external_address(observed_addr);` for `observed_addr` retrieved from `libp2p-identify` exchanges on inbound connections (thus not an ephemeral port).
- Enable `libp2p_tcp::Config::port_reuse`.
- Introduce a `NetworkBehaviour` that ranks the reported addresses and only confirms the most reported ones. Rough draft in https://github.com/libp2p/rust-libp2p/pull/4689.

Reply:
As I mentioned in the other issue, I tried enabling port reuse but it had no effect. I think the first option would complicate our networking code and I would prefer something else unless it was the only option.

Option three sounds the most sensible but based on what Thomas suggested, we could potentially do this filtering in our end as well, not necessarily needing libp2p to provide it. I did a quick and dirty implementation of it and it worked relatively well. If the node is run long enough, a port will be reused and it might get incorrectly confirmed to be a working address so we'd want to filter out these duplicates as well.

Reply:
> As I mentioned in the other issue, I tried enabling port reuse but it had no effect. I think the first option would complicate our networking code and I would prefer something else unless it was the only option.
> 
> Option three sounds the most sensible but based on what Thomas suggested, we could potentially do this filtering in our end as well, not necessarily needing libp2p to provide it. I did a quick and dirty implementation of it and it worked relatively well. If the node is run long enough, a port will be reused and it might get incorrectly confirmed to be a working address so we'd want to filter out these duplicates as well.

Reply:
> As I mentioned in the other issue, I tried enabling port reuse but it had no effect. I think the first option would complicate our networking code and I would prefer something else unless it was the only option.
> 
> Option three sounds the most sensible but based on what Thomas suggested, we could potentially do this filtering in our end as well, not necessarily needing libp2p to provide it. I did a quick and dirty implementation of it and it worked relatively well. If the node is run long enough, a port will be reused and it might get incorrectly confirmed to be a working address so we'd want to filter out these duplicates as well.

You should be able to easily avoid this accidential reuse by keeping an LRU cache only attempt to confirm ~10 addresses or something. All ephemeral ones will drop out of that before the OS reused a port but all stable ports will stay in there.

Reply:
Yes that was the plan but didn't have to time to finish it because something else came up.